rfcs/xxx-v4-query-engine.md

```diff
@@ -0,0 +1,744 @@
+### `findWithCount(params)` => `[Entry[], number]`
+
+Finds and counts entries matching the params
+
+**Params**
+
+| Param    | Type                           | Description                                |
+| -------- | ------------------------------ | ------------------------------------------ |
+| select   | `string[]`                     | Select some attributes to return           |
+| where    | [`WhereParam`](#filtering)     | Filter                                     |
+| limit    | `number`                       | How many entries to return                 |
+| offset   | `number`                       | How many entries to skip                   |
+| orderBy  | [`OrderByParam`](#ordering)    | How to order the entries                   |
+| populate | [`PopulateParam`](#populating) | Select which relations should be populated |
+
+**Example**
+
+```js
+const [entries, count] = await db.query("article").findWithCount({
+  select: ["title", "description"],
+  where: { title: "Hello World" },
+  orderBy: { title: "DESC" },
+  populate: { category: true },
+});
+```
+
+### `create(params)` => `Entry`
+
+Creates one entry and returns it
+
+**Params**
+
+| Param    | Type                           | Description                                |
+| -------- | ------------------------------ | ------------------------------------------ |
+| select   | `string[]`                     | Select some attributes to return           |
+| populate | [`PopulateParam`](#populating) | Select which relations should be populated |
+| data     | `object`                       | Input data                                 |
+
+**Example**
+
+```js
+const entry = await db.query("article").create({
+  data: {
+    title: "My Article",
+  },
+});
+```
+
+### `update(params)` => `Entry`
+
+Updates one entry and returns it
+
+**Params**
+
+| Param    | Type                           | Description                                |
+| -------- | ------------------------------ | ------------------------------------------ |
+| select   | `string[]`                     | Select some attributes to return           |
+| populate | [`PopulateParam`](#populating) | Select which relations should be populated |
+| where    | [`WhereParam`](#filtering)     | Filter                                     |
+| data     | `object`                       | Input data                                 |
+
+**Example**
+
+```js
+const entry = await db.query("article").update({
+  where: { id: 1 },
+  data: {
+    title: "xxx",
+  },
+});
+```
+
+### `delete(params)` => `Entry`
+
+Deletes one entry and returns it
+
+**Params**
+
+| Param    | Type                           | Description                                |
+| -------- | ------------------------------ | ------------------------------------------ |
+| select   | `string[]`                     | Select some attributes to return           |
+| populate | [`PopulateParam`](#populating) | Select which relations should be populated |
```

Comment:
Does `populate` make sense for delete queries? I'm not sure what I'd expect from that ... does it control the `cascade` behavoir?

Reply:
This has the same behaviour create & update to be consistent